### PR TITLE
yak shaving, full deprecation style

### PR DESCRIPTION
**Pull request by Sacha0**
**tl;dr**
This pull request eliminates a few uses of `full` from base/linalg/special.jl, via a minor yak shave. Ref. JuliaLang/LinearAlgebra.jl#229, JuliaLang/LinearAlgebra.jl#231, JuliaLang/julia#18850, and linked threads.

**details**
The eliminated uses of `full` were part of ad hoc checks for upper bidiagonal, lower bidiagonal, and tridiagonal band structure in conversions between structured and/or annotated matrix types.

The first commit generalizes existing and introduces new (not exported) generic functions that perform the necessary upper bidiagonal, lower bidiagonal, and tridiagonal band structure checks. Specifically, the first commit extends `istri[u|l](A)` to `istri[u|l](A, k)` matching `tri[u|l](A, k)`, implements `isbanded(A, kl, ku)` as a short child of `istri[u|l](A, k)`, (re)implements `isdiag(A)`, `isbidiag[u|l](A)`, and `istridiag(A)` as short children of `isbanded(A, ul, uk)`, and tests and documents all those functions.

The second commit eliminates the uses of `full` in question via the functions provided by the first commit.

The third commit then cleans up the complete set of surrounding structured and/or annotated matrix interconversion methods, at a pleasing LOC reduction.

Best!

### CODE REVIEW

File: base/linalg/generic.jl [hunk 1]
```diff
@@ -977,29 +977,36 @@ julia> a = [1 2; 2 -1]
...
 """
-function istriu(A::AbstractMatrix)
+function istriu(A::AbstractMatrix, k::Integer = 0)
     m, n = size(A)
-    for j = 1:min(n,m-1), i = j+1:m
-        if !iszero(A[i,j])
-            return false
+    for j in 1:min(n, m + k - 1)
+        for i in max(1, j - k + 1):m
+            iszero(A[i, j]) || return false
```
Comment (alyst) [inline base/linalg/generic.jl]: maybe `@inbounds`?

**Comment by fredrikekre** [comment]
Looks great, I like the `isbanded` function. I am a bit skeptical to `isbidiag[u|l]` and `istridiag` though. I am not sure they are very useful (seem to be used only a couple of times in this PR, which should indicate something?), it is simple enough to do `isbanded(A, 0, 1)` / `isbanded(A, -1, 0)` / `isbanded(A, -1, 1)`? With introduction of `isbanded` we could perhaps instead `@deprecate isdiag(A) isbanded(A, 0, 0)` rather than introducing more specific functions like that? Thoughts? :smile:

**Comment by StefanKarpinski** [comment]
I think `isdiag` is traditional and common enough that we should keep it. 100% agree that `isbidiag`, `istridiag` etc. should just be expressed using `isbanded`.

**Comment by Sacha0** [comment]
Cheers, agreed :).

File: base/linalg/generic.jl (hunk 1 above)
Comment (Sacha0) [reply base/linalg/generic.jl]: I am not certain this is safe for arbitrary `AbstractArray`s, for example `OffsetArray`s, and so hesitate to add an `@inbounds`?

**Comment by Sacha0** [comment]
Nixed `isbidiag[u|l]` and `istridiag`. Thoughts? Thanks!

**Review by StefanKarpinski — APPROVED** [review]

File: base/linalg/generic.jl (hunk 1 above)
Comment (andreasnoack) [reply base/linalg/generic.jl]: Eventually, these functions should use either be restricted to `StridedMatrix` or the index calculations be changed to work with `OffsetArrays`.

Comment (Sacha0) [reply base/linalg/generic.jl]: Good call :). Do you have thoughts on the best way to generalize these loops? Thanks!

**Comment by Sacha0** [comment]
Thanks all! :)

File: base/linalg/generic.jl (hunk 1 above)
Comment (andreasnoack) [reply base/linalg/generic.jl]: The easiest might be to reuse the existing loops and then use the `i` and `j` to index into the actual indices for the array, e.g. `ii = indices(A, 1)[i]`, and then index with `ii` and `jj`.
